Sources/MongoSwift/Operations/NextOperation.swift

```diff
@@ -17,52 +17,39 @@ internal struct NextOperation<T: Codable>: Operation {
         self.target = target
     }
 
-    // swiftlint:disable:next cyclomatic_complexity
     internal func execute(using _: Connection, session: ClientSession?) throws -> T? {
         // NOTE: this method does not actually use the `connection` parameter passed in. for the moment, it is only
         // here so that `NextOperation` conforms to `Operation`. if we eventually rewrite our cursors to no longer
         // wrap a mongoc cursor then we will use the connection here.
-
         if let session = session, !session.active {
             throw ClientSession.SessionInactiveError
         }
 
-        let out = UnsafeMutablePointer<BSONPointer?>.allocate(capacity: 1)
-        defer {
-            out.deinitialize(count: 1)
-            out.deallocate()
-        }
-
         switch self.target {
         case let .cursor(cursor):
-            // We already check this in `MongoCursor.next()` in order to extract the relevant connection and
-            // session, but error again here just in case.
-            guard case let .open(cursorPtr, _, _, _) = cursor.state else {
-                throw ClosedCursorError
-            }
-            guard mongoc_cursor_next(cursorPtr, out) else {
-                return nil
-            }
+            return try cursor.getNextDocumentFromMongocCursor()
```

Comment:
at this point it feels kind of weird to have this be an operation, but given that we need the logic for calling `mongoc_cursor_next` to be reusable for document caching and for the future `ForEachOperation` and `AllOperation`, I'm not sure how to really improve this.

we could consider just directly calling ` try cursor.getNextDocumentFromMongocCursor()` by using the soon-to-be-added threadpool for method for executing an arbitrary non-operation block instead.

Reply:
I think using the arbitrary non-operation blocks sounds like a decent idea. Once we go pure swift, this weirdness will go away since we'll be able to access the first batch from the `find` command, then only use `getMore` after that.

Reply:
I added a note about this on [SWIFT-668](https://jira.mongodb.org/browse/SWIFT-668) (the ticket where the new `execute` method will be added).